### ISSUE TITLE
Installation issue: icu4c with python3

### ISSUE THREAD

**Issue by iarspider** [spack/spack#14598]
### Steps to reproduce the issue

```console
$ . share/spack/setup-env.sh
$ spack install icu4c
```

### Error Message

```
==> Error: Exception occurred in writer daemon!
Traceback (most recent call last):
  File "/home/sftnight/spack/lib/spack/llnl/util/tty/log.py", line 447, in _writer_daemon
    line = in_pipe.readline()
  File "/usr/lib64/python3.6/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 220: ordinal not in range(128)
==> Error: ProcessError: Command exited with status -13:
    'make' '-j8'
See build log for details:
  /tmp/sftnight/spack-stage/spack-stage-icu4c-65.1-55wm4t2wm3y7ibzmdsn5jwhgqhplfxna/spack-build-out.txt
```

### Information on your system

CentOS7, Python 3.6.8 (taken from the system) - command fails. Python 2.7.5 (also from the system) - command works fine.

Commenting lines 55-58 (i.e. not setting LC_ALL) fixes compilation with Python3 (and doesn't break compilation with Python2).

**Comment by glennpj** [comment]
I am not seeing that behavior here with gcc-9.1.0. The LC_ALL option was put in based on information from the project page:

[For most other platforms, the build\-time environment must be set to a UTF\-8 locale\.
](http://site.icu-project.org/download/59)

Maybe this is more compiler specific. I initially ran into issues with the Intel compiler.

**Comment by glennpj** [comment]
@iarspider Which compiler are you using?

**Comment by glennpj** [comment]
On my system, 
```commandline
$ uname -a
Linux argon-build.hpc 3.10.0-862.14.4.el7.x86_64 #1 SMP Wed Sep 26 15:12:11 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:    CentOS Linux release 7.5.1804 (Core)
```
this package works fine with the LC_ALL code for the following GCC compilers
1. 4.8.5
1. 4.9.4
1. 8.3.0
1. 9.1.0
and the following Intel compilers
1. 17.0.4
1. 18.0.5
1. 19.0.5

The Intel-19 compiler requires the LC_ALL code but the others work with it just fine.

**Comment by iarspider** [comment]
```
$ uname -a
Linux 2bd2d0a6fd97 3.10.0-1062.9.1.el7.x86_64 #1 SMP Fri Dec 6 15:49:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:    CentOS Linux release 7.7.1908 (Core)
$ gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

The build is done inside Docker instance, image: gitlab-registry.cern.ch/sft/docker:lcg-cc7

**Comment by glennpj** [comment]
In the container environment, what is the output of
```commandline
$ locale
```
and
```commandline
$ spack build-env icu4c locale
```

**Comment by iarspider** [comment]
```
$ locale
LANG=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=

$ spack build-env icu4c locale
==> Error: Unable to locate python command in /home/sftnight/spack/opt/spack/linux-centos7-haswell/gcc-4.8.5/python-3.7.6-ul6sonadnjhw2myb4mo3ydxptlsryfel/bin
```

**Comment by glennpj** [comment]
You can try it with a version that does not require python.
```commandline
$ spack build-env icu@57.1 locale
```

**Comment by iarspider** [comment]
```
$ spack build-env icu4c@57.1 locale
LANG=
LC_CTYPE="C"
LC_NUMERIC="C"
LC_TIME="C"
LC_COLLATE="C"
LC_MONETARY="C"
LC_MESSAGES="C"
LC_PAPER="C"
LC_NAME="C"
LC_ADDRESS="C"
LC_TELEPHONE="C"
LC_MEASUREMENT="C"
LC_IDENTIFICATION="C"
LC_ALL=C
```

**Comment by glennpj** [comment]
I wonder if LANG also needs to be set.

**Comment by iarspider** [comment]
Probably related to #15285

**Comment by bernhardkaindl** [comment]
@iarspider: This specific `icu4c` error triggers when spack is executed by python3.5 and python3.6, not with >= 3.7 (likely because later versions of Python have an implicit open(..., encoding='utf-8')), as pointed out in https://github.com/spack/spack/issues/15285#issuecomment-690185837 when spack is called with an effective LC_CTYPE of C or POSIX (or another non-UTF-8 locale). The locale inside the build-env only plays a role if that changes the behavior of the software which is built.

The best fix is likely #21447, which we hopefully should be able to merge following removal of python2.6 as one of the versions on which spack runs. With this info and as this had no activity for more ~18 months and is is duplicate of #15285, I'm closing this.
